### PR TITLE
Feat/KFS-1403 Language service integration

### PR DESCRIPTION
**Pull request by Guttz**
Release Notes
-------------

New Features
- [Early Access] Full Flink SQL autocomplete for `confluent flink shell`

Checklist
---------
- [ ] Leave this box unchecked if features are not yet available in production

What
----

By using the flag `--enable-lsp`, the user can get autocompletions from the new flink language service. The autocompletion experience is now much better and complete. Apart from predicting and suggesting the next keywords possible, we will also complete functions, system properties, statement properties and metadata tokens like catalog, database, table and also columns (both simple and full qualified paths).

References
----------

Test & Review
-------------

Updated shell help golden file
Service itself has over 80% test coverage

### CODE REVIEW

**Comment by brianstrauch** [comment]
> By using the flag `--enable-lsp`

I thought we were going to use a LaunchDarkly flag here? The current method will involve remove the `--enable-lsp` flag when using the LSP is the default functionality, which will be a breaking change.

**Comment by brianstrauch** [comment]
Updated the release notes, let me know if they make sense.

File: .go-version [hunk 1]
```diff
@@ -1 +1 @@
-1.21.4
+1.21.0
```
Comment (brianstrauch) [inline .go-version:1]: Any reason for downgrading?

File: internal/flink/command_shell.go [hunk 2]
```diff
@@ -32,6 +36,7 @@ func (c *command) newShellCommand(prerunner pcmd.PreRunner) *cobra.Command {
 	c.addDatabaseFlag(cmd)
 	pcmd.AddEnvironmentFlag(cmd, c.AuthenticatedCLICommand)
 	pcmd.AddContextFlag(cmd, c.CLICommand)
+	cmd.Flags().Bool("language-service", false, "Enables the flink language service integration (experimental).")
```
Comment (brianstrauch) [inline internal/flink/command_shell.go:39]:
```suggestion
	cmd.Flags().Bool("language-service", false, "Enables the Flink language service integration (experimental).")
```

File: internal/flink/command_shell.go [hunk 3]
```diff
@@ -142,6 +147,20 @@ func (c *command) startFlinkSqlClient(prerunner pcmd.PreRunner, cmd *cobra.Comma
 		return err
 	}
 
+	lspEnabled, err := cmd.Flags().GetBool("language-service")
```
Comment (brianstrauch) [inline internal/flink/command_shell.go:150]:
```suggestion
	languageService, err := cmd.Flags().GetBool("language-service")
```

File: internal/flink/command_shell.go [hunk 4]
```diff
@@ -142,6 +147,20 @@ func (c *command) startFlinkSqlClient(prerunner pcmd.PreRunner, cmd *cobra.Comma
...
+	if err != nil {
+		return err
+	}
+
+	var lspBaseUrl string
+	if lspEnabled {
+		lspBaseUrl, err = c.getFlinkLanguageServiceUrl(flinkGatewayClient)
+		if err != nil {
+			log.CliLogger.Warnf("Shell won't connect to language service. Error getting language service url: %v\n", err)
```
Comment (brianstrauch) [inline internal/flink/command_shell.go:159]:
```suggestion
			log.CliLogger.Warnf("Flink shell failed to connect to language service: error getting language service URL: %v\n", err)
```

File: pkg/flink/app/application_test.go [hunk 5]
```diff
@@ -278,3 +279,86 @@ func (s *ApplicationTestSuite) TestPanicRecovery() {
...
+	}
+	s.inputController.EXPECT().GetUserInput().Times(recoverCount).Do(func() {
+		panic("err in repl")
+	})
+	s.statementController.EXPECT().CleanupStatement().Times(recoverCount)
+
+	// When
+	actual := test.RunAndCaptureSTDOUT(s.T(), func() {
+		run := utils.NewPanicRecovererWithLimit(recoverCount, 3*time.Second)
+		for i := 0; i <= recoverCount; i++ {
```
Comment (brianstrauch) [inline pkg/flink/app/application_test.go:298]: Did you mean to run this `recoverCount + 1` times?
```suggestion
		for i := 0; i < recoverCount; i++ {
```

File: pkg/flink/lsp/completer.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"github.com/sourcegraph/go-lsp"
+
+	"github.com/confluentinc/go-prompt"
+
+	"github.com/confluentinc/cli/v3/pkg/log"
+)
+
+func LSPCompleter(c LSPInterface, configurationSettings func() CliContext) prompt.Completer {
```
Comment (brianstrauch) [inline pkg/flink/lsp/completer.go:11]:
```suggestion
func LspCompleter(c LspInterface, configurationSettings func() CliContext) prompt.Completer {
```

**Review by brianstrauch — COMMENTED** [review]
Other than the comment about the flag that will have to be removed later as a breaking change, all my comments are nits!

**Review by swist — CHANGES_REQUESTED** [review]

File: pkg/flink/app/application_test.go (hunk 5 above)
Comment (Guttz) [reply pkg/flink/app/application_test.go:298]: Yep I did. If I recover 5 times, I have to run it 6 so it panics on the last one

File: .go-version (hunk 1 above)
Comment (Guttz) [reply .go-version:1]: good catch

File: go.mod [hunk 7]
```diff
@@ -1,6 +1,6 @@
 module github.com/confluentinc/cli/v3
 
-go 1.21
+go 1.21.0
```
[the remark this answers is not in the document]
Comment (Guttz) [reply go.mod]: changed

File: pkg/flink/app/.snapshots/TestApplicationTestSuite-TestPanicRecoveryWithLimitWhenLimitExceeded [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Error: internal error. Run `confluent flink shell -vvv` to enable debug logs when starting the flink shell and report the output to the CLI team. Kindly share steps reproduce, if possible.
+Please, restart the CLI.
```
[the remark this answers is not in the document]
Comment (Guttz) [reply pkg/flink/app/.snapshots/TestApplicationTestSuite-TestPanicRecoveryWithLimitWhenLimitExceeded:2]: I don't think it makes sense to use NewErrorWithSuggestions directly here. This is the output from the flink shell where we use the default flink shell "output" helper. We don't return any error to the CLI. But I agree that the format should be the same as the other suggestions.

I can take a look and make sure the output looks similar

[the remark this answers is not in the document]
Comment (Guttz) [reply pkg/flink/app/.snapshots/TestApplicationTestSuite-TestPanicRecoveryWithLimitWhenLimitExceeded:2]: Alright, I've refactored the code so the flinksshell now returns an error. That way, I could use `NewErrorWithSuggestions` directly and the suggestion follows the default CLI format: 


image

File: go.mod [hunk 9]
```diff
@@ -1,6 +1,6 @@
 module github.com/confluentinc/cli/v3
 
-go 1.21
+go 1.21.4
```
Comment (brianstrauch) [inline go.mod:3]: You don't need to include the patch version here. This is updated automatically by service bot
```suggestion
go 1.21
```

Comment (Guttz) [reply go.mod]: Oh, sorry. My newest rebased messed up this. Fixed now

File: internal/flink/command_shell.go [hunk 10]
```diff
@@ -157,10 +175,29 @@ func (c *command) startFlinkSqlClient(prerunner pcmd.PreRunner, cmd *cobra.Comma
...
-	client.StartApp(flinkGatewayClient, c.authenticated(prerunner.Authenticated(c.AuthenticatedCLICommand), cmd, jwtValidator), opts, reportUsage(cmd, c.Config, unsafeTrace))
-	return nil
+	return client.StartApp(flinkGatewayClient, c.authenticated(prerunner.Authenticated(c.AuthenticatedCLICommand), cmd, jwtValidator), opts, reportUsage(cmd, c.Config, unsafeTrace))
+}
+
+func (c *command) getFlinkLanguageServiceUrl(gatewayClient *ccloudv2.FlinkGatewayClient) (string, error) {
+	if cfg := gatewayClient.GetConfig(); cfg != nil && len(cfg.Servers) > 0 {
+		gatewayUrl := cfg.Servers[0].URL
+		parsedUrl, err := url.Parse(gatewayUrl)
+
```
Comment (brianstrauch) [inline internal/flink/command_shell.go:189]:
```suggestion
```

**Review by brianstrauch — APPROVED** [review]

**Review by brianstrauch — APPROVED** [review]

**Pull request merged**